sunpy/net/helioviewer.py

```diff
@@ -171,13 +176,18 @@ def download_jp2(self, date, observatory=None, instrument=None, detector=None,
         >>> hv = helioviewer.HelioviewerClient()  # doctest: +REMOTE_DATA
         >>> filepath = hv.download_jp2('2012/07/03 14:30:00', observatory='SDO',
         ...                            instrument='HMI', detector=None, measurement='continuum')  # doctest: +REMOTE_DATA
+        >>> filepath = hv.download_jp2('2012/07/03 14:30:00', observatory='SDO', measurement='continuum')  # doctest: +REMOTE_DATA
         >>> aia = sunpy.map.Map(filepath)  # doctest: +REMOTE_DATA
         >>> aia.peek()  # doctest: +SKIP
         """
         if source_id is None:
             try:
                 key = (observatory, instrument, detector, measurement)
-                source_id = self.data_sources[key]
+                source_id_list = list(partial_match(key, self.data_sources))
+                if len(source_id_list) == 1:
+                    source_id = source_id_list[0]
+                else:
+                    source_id = self.data_sources[key]
             except KeyError:
                 raise KeyError("The values used for observatory, instrument, detector, measurement "
                                "do not correspond to a source_id. Please check the list using "
```

Comment:
I think since we repeat this entire block of code in two functions, we should move this out into a private function. This way we only need to modify it once.